### ISSUE TITLE
Game Engine

### ISSUE THREAD

**Issue by MonZop** [MonZop/BioBlender#2]
BipoBlender is supposed to use the Blender GE to calculate the transitiono fo a molecule from one conformation to another.
However, as soon as the 'Run In GE' button is hit, Blender crashes

**Comment by zeffii** [comment]
there's an import error, in 2.73, 
```
Blender Game Engine Started
Python module can't be imported - object 'Empty', controller 'Python Script':
ImportError: No module named 'BioBlender2'
Blender Game Engine Finished
```

**Comment by zeffii** [comment]
in 2.73 using the branch (2.73_Comp_Explore) To my surprise the Game engine does run, but doesn't record anything (no cache) probably because of the failed import. I'll investigate.

**Comment by MonZop** [comment]
The recorded animation will be stored as keyframes (one KF for every atom, at each frame), starting 500 frames after the last imported model.
Pls. pay attention that animation makes sense only with at least two different conformations of a protein. If you use file '01_CaM 3 conformations.pdb', and changed nothing in the import, you will have a KF at frame 1 (conformation= model 1), a KF at frame 100 (conf 2), and another one at frame 200. After recording the GE simulation, all frames between 500 and 800 will be keyframed.

**Comment by zeffii** [comment]
![woop](https://cloud.githubusercontent.com/assets/619340/5987603/da464062-a934-11e4-9ccb-398a002e1533.gif)
i'll add some update prints so it shows when a keyframe is added, else it is impossible to tell if it's actually performing useful work

**Comment by zeffii** [comment]
OK i see, those keyframes are inserted upon .pdb import if there are several conformations present in the file.

**Comment by zeffii** [comment]
please forgive what must look like complete ignorance :)

**Comment by MonZop** [comment]
You are kidding? you are giving a very great help, and of course no one is born learnt (italian saying , meaning that one cannot know what has not been tought to her).
It's good that you know programming, and I know biology: only by collaborating we are going to make something new! :)
Now my turn of ignorance: How do I test the new branch?

**Comment by zeffii** [comment]
got pyMOL to generate a surface!! - *jumps up and down*

**Comment by zeffii** [comment]
@MonZop I will start a fresh thread explaining how to test branches:
https://github.com/MonZop/BioBlender/issues/9

**Comment by MonZop** [comment]
When I install Blender 273 it opens and crashes immediately (not BioBlender, the regular Blender). I tried twice (new download, new installation), with the same result...

**Comment by zeffii** [comment]
What I do is have a folder on my C:\ drive called `blender_trunk`, and I drag [builder bot builds](https://builder.blender.org/download/) into it, and have a link on my taskbar pointing at the `blender.exe` of that folder.
![image](https://cloud.githubusercontent.com/assets/619340/6000176/1e13fd46-aadb-11e4-9ce1-12fc5737a577.png)

**Comment by zeffii** [comment]
The dropdown optoins comes from this line
```
split.prop(bpy.context.scene.objects[str(m)].game, "physics_type") 
```
the options are generic to the physics type, from what I can tell:

http://www.blender.org/api/blender_python_api_2_73_5/bpy.types.GameObjectSettings.html?highlight=physics_type#bpy.types.GameObjectSettings.physics_type
(produces the same list)

Perhaps it would be simpler to remove the options and set Rigid Body if the game engine is invoked.

**Comment by zeffii** [comment]
This however does not seem to be the reason for the GE to fail to do anything, it is an `import` issue as mentioned here: https://github.com/MonZop/BioBlender/issues/2#issuecomment-72259622

**Comment by MonZop** [comment]
Yes, if joints are not built, then the recording of GE is not done, as it is meaningless (in biological sense)
As you say, Perhaps it would be simpler to remove the options and set Rigid Body if the game engine is invoked., I agree

**Comment by zeffii** [comment]
could you add a few simple molecules for the purpose of testing the GE , bones / joints ?

**Comment by MonZop** [comment]
OK, let me prepare a file with a very small protein in 4 conformations

**Comment by MonZop** [comment]
Done: you'll find it in the Test Molecules folder. It is a small artificial protein made of only 20 aminoacids, for a total of 158 atoms (excluding Hydrogens).
The file contains 4 MODELS (conformations), but I noticed that BB2 only imports 3 of them. I also tried to import other proteins that have many conformations, but BB2 imports all models except the last one.

**Comment by zeffii** [comment]
OK, when something does 3 of 4 ... of 6 of 7, it's usually an 'off by one' bug, (also known as OBO). will look into that.

**Comment by zeffii** [comment]
This is a perfect example of when to start a fresh Thread, we find a problem, we make a thread. Then the thread can be closed when it gets resolved, or considered not immediately important.

**Comment by MonZop** [comment]
Back to the Game Engine:
In the master_main_first_fixes version of BB, bones are created (they are listed in the Outliner), but the GE is completely still. 
In the Blender console it says:
 `Blender Game Engine Started
Python module can't be imported - object 'Empty', controller 'Python Script':
ImportError: No module named 'BioBlender2'
Blender Game Engine Finished `

**Comment by zeffii** [comment]
There's recent fixes to the BGE, today this: https://developer.blender.org/rBbf157ce9277ba1011a6e317c7629bab0dd8c36a5 it may relate to the problem.

@mikepan can probably make more sense of the BGE stuff (either the Master, or the _experiments_ branch)

**Comment by zeffii** [comment]
![image](https://cloud.githubusercontent.com/assets/619340/6226080/1dbd11e4-b692-11e4-8697-6e021a21cd8e.png)
I renamed BioBlender2.py to BioBlender.py and edited the EmptySet.blend to link to `BioBlender.recorder`, and now the ImportError seems to go away.. Even if this isn't the solution, the naming seems to have caused a problem.

**Comment by zeffii** [comment]
it seems to be doing some thing..
![image](https://cloud.githubusercontent.com/assets/619340/6226210/1dad3b2e-b693-11e4-8f78-d379b6a0bf08.png)

**Comment by MonZop** [comment]
I also just tried, but I get an error:
`Blender Game Engine Started
Python module can't be imported - object 'Empty', controller 'Python Script':
ImportError: No module named 'BioBlender2'`
and no recording of motion

**Comment by MonZop** [comment]
But I am using an old portable, it might be not very updated. I'll check again when I gat back to the office in Pisa (thursday afernoon)

**Comment by zeffii** [comment]
I am using a latest Buildbot version, as it includes Game Engine fixes (which might be good to have anyway)
I've pushed the change to the experimental branch.

further in the right place: https://github.com/MonZop/BioBlender/issues/9#issuecomment-74645953

this will get easier, not harder :)

**Comment by MonZop** [comment]
I have tested with several combinations, but nothing seems to worrk for me: 
![2302getest3](https://cloud.githubusercontent.com/assets/10231334/6328562/0c396edc-bb69-11e4-91f1-91f8cda3e99e.png)
any clue?

**Comment by zeffii** [comment]
I think this happens if the Addon folder is not called _BioBlender_, which might be the case if you are using the zip. You're might be called BioBlender-master , or one of the branches. Ideally the addon wouldn't care about the name of the folder inside which it lives.

**Comment by zeffii** [comment]
So if you are trying the version i'm experimenting with that folder would be called:

 `BioBlender-master_mff_experiments` , but rename it to `BioBlender`

the content of this version should be: 
![image](https://cloud.githubusercontent.com/assets/619340/6329082/929ac4a4-bb6d-11e4-930f-35b675049110.png)

If any other BioBlender addons are enabled / ticked other than this one  untick them, save settings / Ctrl+U to store this preference. Restart Blender. When you navigate to the addons panel again, you can see where it's loading the addon from by viewing the dropdown:

![image](https://cloud.githubusercontent.com/assets/619340/6329152/2418d1c8-bb6e-11e4-98a7-c7ee332f0b05.png)

**Comment by zeffii** [comment]
Now i said: 
>  Ideally the addon wouldn't care about the name of the folder inside which it lives.

Most likely it can be made to be entirely agnostic about folder names, but that's a different problem (one I would like to solve soon)

**Comment by MonZop** [comment]
Oh, finally, I managed! :-)
![worikngge](https://cloud.githubusercontent.com/assets/10231334/6330058/9ec2f4ba-bb75-11e4-93c0-6885b8e2dd26.png)
